pandas/indexes/base.py

```diff
@@ -199,14 +199,25 @@ def __new__(cls, data=None, dtype=None, copy=False, name=None,
                             data = np.array(data, copy=copy, dtype=dtype)
                         elif inferred in ['floating', 'mixed-integer-float']:
 
-                            # if we are actually all equal to integers
+                            # If we are actually all equal to integers,
                             # then coerce to integer
-                            from .numeric import Int64Index, Float64Index
+                            from .numeric import (Int64Index, UInt64Index,
+                                                  Float64Index)
                             try:
-                                res = data.astype('i8')
+                                res = data.astype('i8', copy=False)
                                 if (res == data).all():
                                     return Int64Index(res, copy=copy,
                                                       name=name)
+                            except (OverflowError, TypeError, ValueError):
+                                pass
```

Comment:
shouldn't we try to convert to a UInt64 index *only* if we have oveflow on the int64 conversion? (and not some other error)? 

I suppose the way the code is now it will just hit both paths then fall thru on a genuine Type/Value error.